components/salsa-2022-macros/src/salsa_struct.rs

```diff
@@ -25,41 +25,46 @@
 //! * data method `impl Foo { fn data(&self, db: &dyn crate::Db) -> FooData { FooData { f: self.f(db), ... } } }`
 //!     * this could be optimized, particularly for interned fields
 
-use heck::ToUpperCamelCase;
-use proc_macro2::{Ident, Literal, Span, TokenStream};
-use syn::spanned::Spanned;
-
+use crate::modes::Mode;
 use crate::{
     configuration,
+    modes::AllowedModes,
     options::{AllowedOptions, Options},
 };
+use heck::ToUpperCamelCase;
+use proc_macro2::{Ident, Literal, Span, TokenStream};
+use syn::spanned::Spanned;
 
-pub(crate) struct SalsaStruct<A: AllowedOptions> {
+pub(crate) struct SalsaStruct<A: AllowedOptions, M: AllowedModes> {
```

Comment:
You can get away with not adding an extra bound if you do something like this:
```suggestion
pub(crate) struct SalsaStruct<A: AllowedOptions + AllowedModes> {
```

Though, I don't think AllowedModes is necessary right now, since it's already rejected by option parsing before we even get to `check_singleton`

Reply:
Oh yeah. True. Thanks!

About the second part, It's not exactly rejected by option parsing. Option parsing still allows any salsa struct to be `singleton`. I tried doing the validation there, but apparently the `syn::parse::ParseStream` supplied to the `stn::parse` trait only contains the args to the macro.

Reply:
I'm a bit confused by this. The option parsing for the struct itself would be e.g. to parse things like `#[salsa::input(jar = Foo)]`, so it seems like it ought to also be usable to parse the singleton flag?

Reply:
I'm the confused one. Will work on it. Thanks

Reply:
@DropDemBits  I was wrong. I could just use `AllowedOptions` and do away with `AllowedModes` altogether. 
Thanks!